Improve subproc autotuning implementation

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #149890
* __->__ #149700

Summary: The primary change is to update the autotune-in-a-subproc implementation to avoid using multiprocessing spawn. Spawn (re)executes the toplevel script in the subproc, which can be problematic. The approach here is similar to Triton parallel compile: we Popen a subproc on a controlled entry point and communicate over pipes. That change drove a lot of refactoring in the TuningProcess class, so I took the opportunity to simplify some things, rename some methods, etc. 

One other notable change is around the timeout / kill approach. After a timeout, we were previously attempting to stop the subproc in three steps (graceful shutdown, sigkill if graceful fails, sigterm if sigkill fails). I'm gonna argue think that's not useful: 1) The graceful shutdown is never going to work unless the subproc happens to have just completed its task and is ready to receive the next command. 2) If we're going to kill the subproc, let's just take the most aggressive approach and move on as quickly as possible to restarting it rather than waiting to see if previous shutdown attempts succeeded. The only downside that I can find find is maybe a little log spew?, e.g., ` ResourceWarning: subprocess 2987680 is still running`

List of changes:
* Use Popen instead of spawn for the autotuning subprocess.
* Introduced a new entry point `__autotune_main__.py`
* Renamed some TuningProcess methods. For example `shutdown` makes more sense than `terminate` because the latter implies a forced kill.
* Simplified the implementation around benchmarking timeout and how we kill the subproc after a timeout.
* Deprecated the unused timeout configs in `_inductor/config.py`
* Moved `get_ld_library_path` helper to a common utils file.
* Added more unit tests for subproc crashes / timeouts / exceptions, etc.

Test plan:
* New unit tests
* Also ran internally with all combinations of: build mode `opt` and `dev-nosan`, and `buck run` vs. executing the `.par` file directly.
* Made sure the functionality to parallelize autotuning across different GPUs is working (it wasn't clear to me this was behaving the way we wanted it to).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

Differential Revision: [D71976971](https://our.internmc.facebook.com/intern/diff/D71976971)

masnesral committed:
[WIP] Improve subproc autotuning

[ghstack-poisoned]

masnesral committed:
Update on "[WIP] Improve subproc autotuning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

masnesral committed:
Update on "[WIP] Improve subproc autotuning"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

masnesral commented on test/inductor/test_max_autotune.py:
```diff
@@ -1235,63 +1208,134 @@ def __str__(self) -> str:
+        p.shutdown()
+
+
+class TestTuningProcessPool(TestCase):
```
There aren't as many changes to the unit tests below as it appears in the diff. Mostly I just added a new test to exercise a timeout (and make sure the subprocess is still usable for another job after a timeout).

masnesral commented on torch/_inductor/__autotune_main__.py:
```diff
@@ -0,0 +1,33 @@
+    write_pipe = os.fdopen(args.write_fd, "wb")
+
+    try:
+        # Ensures the subprocess exits if the parent crashes:
```
Reused this handy function from the Triton parallel compile impl. Could move it so some common utils class if anyone feels strongly.

masnesral replied:
@eellison, I'm pretty much taking the comment for granted that it does what it says it does: https://github.com/pytorch/pytorch/blob/5c19952c8342a854fcb0ca5714f1e40da199f94c/torch/_inductor/compile_worker/utils.py#L16-L24

masnesral commented on torch/_inductor/autotune_process.py:
```diff
@@ -1,23 +1,26 @@
 from __future__ import annotations
 
-import contextlib
+import atexit
```
Apologies for the size of the changes in this file. I'd suggest just reviewing the new implementation of the TuningProcess by itself instead of trying to grok the changes. The diff isn't lining up that well and the TuningProcess impl by itself is actually not that big.

masnesral commented on torch/_inductor/autotune_process.py:
```diff
@@ -49,245 +50,193 @@
-log = logging.getLogger(__name__)
-
-
-# Used to synchronize between parent and child processes
```
I don't know of any reason to perform this handshake when the subproc starts.

masnesral commented on torch/_inductor/autotune_process.py:
```diff
@@ -49,245 +50,193 @@
     pass
 
 
-@contextlib.contextmanager
```
Not needed because the new implementation doesn't need to temporarily change the environment in the parent process. We can initialize the environment of the child directly when we Popen and set the right value for CUDA_VISIBLE_DEVICES. For those that aren't familiar with what's happening here: we have functionality to start multiple autotuning subprocs, each using a different GPU (for autotuning parallelism). I don't think it's widely used, but maybe we can enable it more widely now?

masnesral commented on torch/_inductor/autotune_process.py:
```diff
@@ -49,245 +50,193 @@
-        assert self.request_queue is not None
-        self.request_queue.put(obj)
-
-    def get(
```
@kadeng I think you added the timeout stuff? If so, wdyt? See my comments in the PR summary. I don't think we get any value out of attempting to kill shutdown the subproc gracefully before killing it forcefully. So I've simplified the treatment. Please LMK if you think I've missed something.

masnesral commented on torch/_inductor/autotune_process.py:
```diff
@@ -298,34 +247,19 @@ def initialize(self) -> None:
         # the number of threads with the number of devices.
         self.executor = ThreadPoolExecutor(max_workers=len(devices))
 
-        # Register the exit handler for the parent process so it will terminate
```
I'm responsible for this part during my hackamonth, but it's weird. Why have a single global registration inside an instance method? This class is supposed to be a singleton effectively. I moved the (single) exit registration next to where we create that singleton below.

masnesral commented on torch/_inductor/autotune_process.py:
```diff
@@ -563,21 +503,38 @@ def benchmark(
     """
-    Supports unit testing. Defined in this file so that the TuningProcess
-    sub-process knows how to unpickle these objects.
+    Supports unit testing. Defined in this file instead of the test file so the
```
Hopefully this explains it. This class facilitates testing by allowing me to cause the autotune subproc to do various useful things (like crash itself). Morally this class wants to be declared in the unit test file, since that's where it's used. But because pickle doesn't serialize code, our autotune subproc can't pickle this class if it doesn't know the definition.

aorenste commented on torch/_inductor/autotune_process.py:
```diff
@@ -49,245 +50,193 @@
-    processes: Optional[queue.Queue[TuningProcess]] = None
-    executor: Optional[ThreadPoolExecutor] = None
+    def __init__(self) -> None:
+        self.processes: Optional[queue.Queue[TuningProcess]] = None
+        self.executor: Optional[ThreadPoolExecutor] = None
```
Maybe these should be non-None so we don't need to assert they're not None below?

masnesral replied:
Yeah, will clean up this part in a follow-up PR

aorenste commented on torch/_inductor/autotune_process.py:
```diff
@@ -346,17 +280,17 @@ def get_device_list(self) -> Sequence[Optional[int]]:
             self.executor = None
 
         if self.processes is not None:
             for p in self.processes.queue:
```
Do we need to worry about processes that were taken out of the queue because they're currently benchmarking?

aorenste commented on torch/_inductor/autotune_process.py:
```diff
@@ -371,19 +305,24 @@ def target(self, choice: TritonTemplateCaller) -> float:
         assert self.processes is not None
 
         process = self.processes.get()
```
Do we need to worry about matching the GPU to the kernel being tested? i.e. if I have a machine with both an A100 and an H100 and I'm making a kernel for the H100 it's going to be a problem if I pull an A100 out of the queue, won't it?

aorenste reviewed (approved):
Looks generally reasonable to me. A few questions/comments.

masnesral commented on torch/_inductor/autotune_process.py:
```diff
@@ -49,245 +50,193 @@
+        if self.alive():
+            TuningProcess.send(None, self.write_pipe)
+        if wait:
+            self.wait()
```
Yes we can call it twice, but it looks like the second is a no-op (according to LLM advice and a real one-off experiment)

masnesral committed:
Update on "Improve subproc autotuning implementation"


Summary: The primary change is to update the autotune-in-a-subproc implementation to avoid using multiprocessing spawn. Spawn (re)executes the toplevel script in the subproc, which can be problematic. The approach here is similar to Triton parallel compile: we Popen a subproc on a controlled entry point and communicate over pipes. That change drove a lot of refactoring in the TuningProcess class, so I took the opportunity to simplify some things, rename some methods, etc. 

One other notable change is around the timeout / kill approach. After a timeout, we were previously attempting to stop the subproc in three steps (graceful shutdown, sigkill if graceful fails, sigterm if sigkill fails). I'm gonna argue think that's not useful: 1) The graceful shutdown is never going to work unless the subproc happens to have just completed its task and is ready to receive the next command. 2) If we're going to kill the subproc, let's just take the most aggressive approach and move on as quickly as possible to restarting it rather than waiting to see if previous shutdown attempts succeeded. The only downside that I can find find is maybe a little log spew?, e.g., ` ResourceWarning: subprocess 2987680 is still running`

List of changes:
* Use Popen instead of spawn for the autotuning subprocess.
* Introduced a new entry point `__autotune_main__.py`
* Renamed some TuningProcess methods. For example `shutdown` makes more sense than `terminate` because the latter implies a forced kill.
* Simplified the implementation around benchmarking timeout and how we kill the subproc after a timeout.
* Deprecated the unused timeout configs in `_inductor/config.py`
* Moved `get_ld_library_path` helper to a common utils file.
* Added more unit tests for subproc crashes / timeouts / exceptions, etc.

Test plan:
* New unit tests
* Also ran internally with all combinations of: build mode `opt` and `dev-nosan`, and `buck run` vs. executing the `.par` file directly.
* Made sure the functionality to parallelize autotuning across different GPUs is working (it wasn't clear to me this was behaving the way we wanted it to).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

yangw-dev commented:
pytorchbot revert -m="This is breaking servicelab_benchmark_pyper_local_runner internally" -c=ghfirst

yangw-dev commented:
@pytorchbot revert -m="This is breaking servicelab_benchmark_pyper_local_runner internally" -c=ghfirst

masnesral committed:
Update on "Improve subproc autotuning implementation"


Summary: The primary change is to update the autotune-in-a-subproc implementation to avoid using multiprocessing spawn. Spawn (re)executes the toplevel script in the subproc, which can be problematic. The approach here is similar to Triton parallel compile: we Popen a subproc on a controlled entry point and communicate over pipes. That change drove a lot of refactoring in the TuningProcess class, so I took the opportunity to simplify some things, rename some methods, etc. 

One other notable change is around the timeout / kill approach. After a timeout, we were previously attempting to stop the subproc in three steps (graceful shutdown, sigkill if graceful fails, sigterm if sigkill fails). I'm gonna argue think that's not useful: 1) The graceful shutdown is never going to work unless the subproc happens to have just completed its task and is ready to receive the next command. 2) If we're going to kill the subproc, let's just take the most aggressive approach and move on as quickly as possible to restarting it rather than waiting to see if previous shutdown attempts succeeded. The only downside that I can find find is maybe a little log spew?, e.g., ` ResourceWarning: subprocess 2987680 is still running`

List of changes:
* Use Popen instead of spawn for the autotuning subprocess.
* Introduced a new entry point `__autotune_main__.py`
* Renamed some TuningProcess methods. For example `shutdown` makes more sense than `terminate` because the latter implies a forced kill.
* Simplified the implementation around benchmarking timeout and how we kill the subproc after a timeout.
* Deprecated the unused timeout configs in `_inductor/config.py`
* Moved `get_ld_library_path` helper to a common utils file.
* Added more unit tests for subproc crashes / timeouts / exceptions, etc.

Test plan:
* New unit tests
* Also ran internally with all combinations of: build mode `opt` and `dev-nosan`, and `buck run` vs. executing the `.par` file directly.
* Made sure the functionality to parallelize autotuning across different GPUs is working (it wasn't clear to me this was behaving the way we wanted it to).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71976971).

masnesral commented:
@pytorchbot merge

masnesral committed:
Update on "Improve subproc autotuning implementation"


Summary: The primary change is to update the autotune-in-a-subproc implementation to avoid using multiprocessing spawn. Spawn (re)executes the toplevel script in the subproc, which can be problematic. The approach here is similar to Triton parallel compile: we Popen a subproc on a controlled entry point and communicate over pipes. That change drove a lot of refactoring in the TuningProcess class, so I took the opportunity to simplify some things, rename some methods, etc. 

One other notable change is around the timeout / kill approach. After a timeout, we were previously attempting to stop the subproc in three steps (graceful shutdown, sigkill if graceful fails, sigterm if sigkill fails). I'm gonna argue think that's not useful: 1) The graceful shutdown is never going to work unless the subproc happens to have just completed its task and is ready to receive the next command. 2) If we're going to kill the subproc, let's just take the most aggressive approach and move on as quickly as possible to restarting it rather than waiting to see if previous shutdown attempts succeeded. The only downside that I can find find is maybe a little log spew?, e.g., ` ResourceWarning: subprocess 2987680 is still running`

List of changes:
* Use Popen instead of spawn for the autotuning subprocess.
* Introduced a new entry point `__autotune_main__.py`
* Renamed some TuningProcess methods. For example `shutdown` makes more sense than `terminate` because the latter implies a forced kill.
* Simplified the implementation around benchmarking timeout and how we kill the subproc after a timeout.
* Deprecated the unused timeout configs in `_inductor/config.py`
* Moved `get_ld_library_path` helper to a common utils file.
* Added more unit tests for subproc crashes / timeouts / exceptions, etc.

Test plan:
* New unit tests
* Also ran internally with all combinations of: build mode `opt` and `dev-nosan`, and `buck run` vs. executing the `.par` file directly.
* Made sure the functionality to parallelize autotuning across different GPUs is working (it wasn't clear to me this was behaving the way we wanted it to).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D71976971](https://our.internmc.facebook.com/intern/diff/D71976971)

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

clee2000 commented:
Made no changes and reissued merge command, I think the check was just a bit out of date
